gatereader/reader.py

```diff
@@ -0,0 +1,187 @@
+import serial
+import struct
+from enum import Enum
+from .utils.structparse import mystruct, t, STRUCT_FORMAT
+
+class PacketId(Enum):
+    INVALID_PACKET        = 0x00
+    ANSWER_TO_RESET       = 0x01
+    CONTINUE_BOOT         = 0x02
+    FIRMWARE_UPGRADE      = 0x03
+    FW_UPGRADE_READY      = 0x04
+    FW_WRITE              = 0x05
+    FW_WRITE_ACK          = 0x06
+    FW_UPDATE_FINISH      = 0x07
+    GET_STATUS            = 0x08
+    SET_LED               = 0x09
+    ACK                   = 0x0A
+    BEEP                  = 0x0B
+    RFID_SEND             = 0x0C
+    RFID_SEND_COMPLETE    = 0x0D
+    RX_ERROR              = 0xFF
+
+class ResponseLength(Enum):
+    RESPONSE_ATR          = 5
+    TRANSMIT_OVERHEAD     = 3
+
+class Reader:
+
+    class ResetException(Exception):
+        """The reader has been unexpectedly reset."""
+        pass
+
+    class ReaderError(Exception):
+        """The reader did something wrong and MUST be reset."""
+        pass
+
+    class CorruptedPacketException(Exception):
+        pass
+
+    class Leds(Enum):
+        RED_LED     = 0
+        GREEN_LED   = 1
+        BLUE_LED    = 2
+
+    BAUDRATE   = 19200
+    BYTESIZE   = serial.EIGHTBITS
+    PARITY     = serial.PARITY_EVEN
+    STOPBITS   = serial.STOPBITS_ONE
+    RX_TIMEOUT = 0.5
+
+    MAX_BEEP_LENGTH = 8
+    MAX_RFID_PAYLOAD = 128
+
+    PacketHead = mystruct('PacketHead',
```

Comment:
Note that `mystruct` has changed considerably and you will have to do a bit of rewriting when I split it into its own package that you'll depend on -- just FYI. E.g. the tail is now part of the struct (if you declare it to be a struct with tail).